workspaces/vscode-python/Dockerfile

```diff
@@ -0,0 +1,84 @@
+FROM codercom/code-server:4.22.0-ubuntu
+
+# On PL, we want to standardize on using 1001:1001 for the user. We change the
+# "coder" account's UID and GID here so that fixuid has no work to do later.
+# This speeds up container loading time drastically and avoids timeouts.
+USER root
+RUN OLD_UID=$(id -u coder) && \
+    OLD_GID=$(id -g coder) && \
+    NEW_UID=1001 && \
+    NEW_GID=1001 && \
+    groupmod -g $NEW_GID coder && \
+    usermod -u $NEW_UID -g $NEW_GID coder && \
+    find /home -user $OLD_UID -execdir chown -h $NEW_UID {} + && \
+    find /home -group $OLD_GID -execdir chgrp -h $NEW_GID {} +
+
+# Remove sudo permissions for default user, then test
+USER root
+RUN find /etc/sudoers.d -type f -not -name README -delete && \
+    sudo -u coder bash -c 'if sudo echo "Test failed: User can sudo" ; then false ; fi'
+
+# Apt installs: Enable manpages with "unminimize", and then install common utilities.
+USER root
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+    yes | unminimize && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    find /tmp -not -path /tmp -delete
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+    curl git htop less nano unzip vim wget zip && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    find /tmp -not -path /tmp -delete
+
+# Install Python and required packages via miniconda
+USER coder
+COPY requirements.txt /
+RUN export arch=`uname -m` \
+ && curl -sfLO "https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-${arch}.sh" \
+ && chmod +x "Miniconda3-latest-Linux-${arch}.sh" \
+ && ./"Miniconda3-latest-Linux-${arch}.sh" -b -p /home/coder/miniconda \
+ && PATH="/home/coder/miniconda/bin:$PATH" \
+ && /home/coder/miniconda/condabin/conda init bash \
+ && conda config --add channels conda-forge \
+ && conda run -n base pip install -r /requirements.txt \
+ && rm "Miniconda3-latest-Linux-${arch}.sh" \
+ && conda clean --all
+
+# Install VS Code extensions and clear the extension cache to reduce image size.
+USER coder
+RUN code-server --disable-telemetry --force \
+    # vscode support for python, including debugger
+    --install-extension ms-python.python \
+    # auto-generate templates for function docstrings
+    --install-extension njpwerner.autodocstring \
+    # auto-fix indentation for multiline contexts
+    --install-extension KevinRose.vsc-python-indent \
+    # visualize csv files in vscode window
+    --install-extension mechatroner.rainbow-csv \
+    # linting for markdown files
+    --install-extension DavidAnson.vscode-markdownlint \
+    # format tables in markdown files
+    --install-extension fcrespo82.markdown-table-formatter \
+    && rm -rf /home/coder/.local/share/code-server/CachedExtensionVSIXs
```

Comment:
Keeping in mind the possibility that a future PR would split the base and Python-specific content into separate images, would it make sense to be prepared for that and move all Python-specific commands to the bottom of the Dockerfile?

Reply:
I'm fine keeping the order largely how it is, though you might consider moving extension installation above the `requirements.txt` stuff for better layer caching/reuse?

Reply:
I thought there was some reason why the Miniconda Python stuff needed to be installed before the extension installs would complete successfully, but I haven't tested that. (I thought one of the others mentioned this to me, but I could be misremembering.)

Reply:
I think if that's the case, we should document why the order matters, I often find it helpful to leave comments about when ordering matters implicitly.

Reply:
@firasm I could try to build this with the order inverted, but from your memory, does the installation of the Python extensions require the Conda installs to complete first? (If I try it and it _seems_ to work, I'm not sure if it will actually be fully working.) Or we can add a comment to document this as Nathan says.

Reply:
The Python extension from Microsoft clearly states that a supported version of python should be installed first so we should just follow that guidance and add a comment 

![Screenshot 2024-05-30 at 1 08 29 PM](https://github.com/PrairieLearn/PrairieLearn/assets/2507459/d291388e-0758-4e0d-b024-ff144a8fe715)

Of course, nothing to say it WON'T work if the order is inverted, but why play with fire 🙃

Reply:
I put comments in the Dockerfile to this effect now.

Reply:
It's worth noting here (at least) that the extension gets installed from the Open VSX marketplace, not from the MS one, but the extension is in fact a republished version of Microsoft's extension. Open VSX did that within the terms of the license, and MS also commented on their GitHub repo that it was okay to do so with a custom-built vsix.